services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/RouterServiceImpl.java

```diff
@@ -0,0 +1,316 @@
+            return null;
+        }
+        for (RouteTable routeTable : vpcRouteTable) {
+            String routeTableType = routeTable.getRouteTableType().getRouteTableType();
+            if (RouteTableType.PRIVATE_SUBNET.getRouteTableType().equals(routeTableType) || RouteTableType.PUBLIC_SUBNET.getRouteTableType().equals(routeTableType)) {
+                throw new VpcRouterContainsSubnetRoutingTables();
+            }
+        }
+
+        // delete router and route tables
+        this.routerDatabaseService.deleteRouter(router.getId());
+        for (RouteTable routeTable : vpcRouteTable) {
+            this.routeTableDatabaseService.deleteRouteTable(routeTable.getId());
+        }
+
+        return router.getId();
+    }
+
+    @Override
+    public RouteTable getOrCreateVpcRouteTable(String projectId, String vpcId) throws DatabasePersistenceException, CanNotFindVpc, CacheException, ExistMultipleVpcRouter {
+        RouteTable routeTable = null;
+
+        // Get or create a router for a Vpc
+        Router router = getOrCreateVpcRouter(projectId, vpcId);
+
+        // If VPC has a VPC routing table, return the routing table’s state
+        List<RouteTable> vpcRouteTables = router.getVpcRouteTable();
+        for (RouteTable vpcRouteTable : vpcRouteTables) {
+            String routeTableType = vpcRouteTable.getRouteTableType().getRouteTableType();
+            if (RouteTableType.VPC.getRouteTableType().equals(routeTableType)) {
+                return vpcRouteTable;
+            }
+        }
+
+        // If VPC doesn’t have a VPC routing table, this operation will create a VPC routing table and pump-in the VPC default routing rules.
+        routeTable = createDefaultVpcRouteTable(projectId, router);
+
+        return routeTable;
+    }
+
+    @Override
+    public RouteTable createDefaultVpcRouteTable(String projectId, Router router) throws DatabasePersistenceException {
+        String routeTableId = UUID.randomUUID().toString();
+        String routeEntryId = UUID.randomUUID().toString();
+        String owner = router.getOwner();
+        List<RouteTable> vpcRouteTables = router.getVpcRouteTable();
+        List<RouteEntry> routeEntities = new ArrayList<>();
+
+        // create a VPC routing table and pump-in the VPC default routing rules
+        RouteEntry routeEntry = new RouteEntry(projectId, routeEntryId, "default_vpc_routeEntry", "", null, RouteConstant.DEFAULT_TARGET, RouteConstant.DEFAULT_PRIORITY, routeTableId, null);
+        routeEntities.add(routeEntry);
+        this.routeEntryDatabaseService.addRouteEntry(routeEntry);
+
+        RouteTable routeTable = new RouteTable(projectId, routeTableId, "default_vpc_routeTable", "", routeEntities, RouteTableType.VPC, owner);
+        vpcRouteTables.add(routeTable);
+        this.routeTableDatabaseService.addRouteTable(routeTable);
+
+        vpcRouteTables.add(routeTable);
+        router.setVpcRouteTable(vpcRouteTables);
+        this.routerDatabaseService.addRouter(router);
```

Comment:
These lines of code same as the lines in the `getOrCreateVpcRouter` function. Since one VPC has only one router and default routing table. Should we merge `createDefaultVpcRouteTable` into `getOrCreateVpcRouter` ?

Reply:
Although some codes here are same as createDefaultVpcRouter, it is not convenient to merge them. They are used for different method and the parameters passed in are dfferent, either